### ISSUE TITLE
remove-unused-components skips schemas that have a top-level `allOf`

### ISSUE THREAD

**Issue by jedelson-pagerduty** [Redocly/redocly-cli#1209]
**Describe the bug**

Given an OpenAPI document that contains something along these lines:

```
openapi: "3.0.0"
paths:
  /pets:
    get:
      summary: List all pets
      operationId: listPets
      responses:
        '200':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Used'
components:
  schemas:
    Unused:
      allOf:
        - type: object
          properties:
            one:
              type: string
        - type: object
          properties:
            two:
              type: string
    Used:
      type: object
      properties:
        link:
          type: string
        child:
          $ref: '#/components/schemas/Two'
```

It would be expected that bundle --remove-unused-components would remove the Unused schema since it isn't used.

However, the current behavior is that after running bundle with --removed-unused-components, Unused is not removed. This _appears_ to be intentional (see https://github.com/Redocly/redocly-cli/blob/cfa31330b04f22e7b855143b2eef0d040b0873fa/packages/core/src/rules/oas3/remove-unused-components.ts#L64) but I have not found any reference to _why_ it was intentional.

**To Reproduce**

1. Copy the OpenAPI content above to a file.
2. Run redocly bundle <file.yaml> -o output.yaml --remove-unused-components
3. Check the output.yaml file.

**Expected behavior**

An unused component gets removed regardless of whether it has an `allOf` or not.

**OpenAPI definition**

Have only tested with 3.0 but the same `if` statement is present in the `oas2` rule.

**Redocly Version(s)**

1.0.0 and main.

**`Node.js` Version(s)**

16.17.0, 16.18.0, and 18.16.0

**Comment by jedelson-pagerduty** [comment]
@slavikbez I think this came from your PR (#443)

**Comment by tatomyr** [comment]
@jedelson-pagerduty I believe those changes mirror this rule: https://github.com/Redocly/redocly-cli/blob/572980cb33e3ff399e035246e8aa2c24e54d9617/packages/core/src/rules/oas3/no-unused-components.ts#L43
Looks like this is a workaround to detect discriminators.

**Comment by tatomyr** [comment]
I haven't found any related tests that can shed some light on what was the case for this 😄.
@RomanHotsiy could you maybe help with an example of a false positive?

**Comment by RomanHotsiy** [comment]
Yes, as @jedelson-pagerduty noted in the updated comment, the problem is related to implicit discriminator mapping.

Someone can point to `Pet` only and the nested schemas are supposed to be kept too. This requires a careful implementation. There are many edge-cases to this.

**Comment by jokull** [comment]
I think I'm hitting this issue with the the Cloudflare schema. Try tree shaking this schema for example: https://gist.github.com/jokull/2ccf3a492fd6fe35e2b7cd07704f5863

**Comment by VladCananau** [comment]
I am encountering this issue as well in a codebase that makes heavy use of allOf. 

Any chance for a bump in priority ? Cheers

**Comment by tatomyr** [comment]
@VladCananau unfortunately, I cannot put any estimates on this. There was already an attempt to solve this and we discussed the approach, but the author disappeared.

**Comment by jokull** [comment]
I [hacked this](https://gist.github.com/jokull/0db70eb2ac8ab12e4ad1742907220135) together using [spotlight](https://stoplight.io)

**Comment by LaSylv** [comment]
I managed to make something work, but using a preprocessor as you need more context than given in the visitors to find who is used/used by unused only, etc...

It looks like this https://gist.github.com/LaSylv/3afff365332360015723457a61d71962


@tatomyr Would you be open to a PR in this direction ? If so, I can work to create the PR, clean the code, make it to typescript and add unit tests

**Comment by tatomyr** [comment]
@LaSylv sure you can try, but I don't think you need a preprocessor for that. Maybe you'll have to use the `enter` visitor instead of `leave`? Or both? You can prepare a draft PR and we'll take a look at it. Please also scroll through the comments to the previous attempt to solve this: https://github.com/Redocly/redocly-cli/pull/1220 for additional context.

**Comment by tjwoodall** [comment]
I've also hit this in the case where I'm also using --dereference so I can be sure none of the schemas can be being used.

**Comment by tatomyr** [comment]
Released in v2.31.2.
Thanks to @jedelson-pagerduty for the issue and the original solution!
